### PR TITLE
Add latest time zero-order-hold sync policy

### PR DESCRIPTION
**Pull request by andermi**
Rate logic added to #38.
Computes rates of received messages and pivots on fastest topic; essentially, upsamples slower topics with a zero-order hold.

### CODE REVIEW

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a few comments inline.

Besides that, it might be nice to add in some documentation to the top of `latest_time.h` that describes briefly what this sync policy does, and how to use it.

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h]: removed

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h]: alphabetized, and removed unused

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp]: removed

**Comment by andermi** [comment]
Thanks @clalancette ! I added some some documentation at the top of the header.

File: test/test_latest_time_policy.cpp
Comment (chapulina) [inline test/test_latest_time_policy.cpp:132]: I see that other tests in this repository make heavy use of asserts, but I'd recommend using `EXPECT`s in places where a failure won't cause the test to crash. This way when the test does fail, it still continues and can have other failures too. That can be useful to find out the root cause of the failure, because there's more information.

File: test/test_latest_time_policy.cpp
Comment (chapulina) [inline test/test_latest_time_policy.cpp:171]: Both tests are the same up to here. Consider moving the common code to a function. Maybe even to `LatestTime::SetUp` so it's called at the beginning of every test.

File: test/test_latest_time_policy.cpp
Comment (chapulina) [inline test/test_latest_time_policy.cpp:186]: Consider condensing this entire switch block into something like this

```
    if (idx > 2U)
    {
      ASSERT_EQ(h.count_, idx - 2);
      ASSERT_EQ(h.p_->data, p[idx]->data);
      ASSERT_EQ(h.q_->data, q[(idx - 1) / 2]->data);
      ASSERT_EQ(h.r_->data, r[(idx - 3) / 4]->data);
    }
    else
    {
      ASSERT_FALSE(h.p_);
      ASSERT_FALSE(h.q_);
      ASSERT_FALSE(h.r_);
    }
```

File: include/message_filters/sync_policies/latest_time.h
Comment (chapulina) [inline include/message_filters/sync_policies/latest_time.h:215]: I'm trying to understand the motivation for resetting every 10k calls. Maybe it would be smoother to use a moving window and always look back at the Nth previous call?

Also, maybe `compute_hz` could be a member of `Rate`, but you'd have to pass in the current time...

File: include/message_filters/sync_policies/latest_time.h
Comment (chapulina) [inline include/message_filters/sync_policies/latest_time.h:125]: I don't see these tuples being used, are they needed?

File: include/message_filters/sync_policies/latest_time.h
Comment (chapulina) [inline include/message_filters/sync_policies/latest_time.h:272]: Is this supposed to be used somewhere?

**Review by chapulina — COMMENTED** [review]
The functionality works for me. I have some questions and suggestions. Take it all with a grain of salt since it's my first time touching this repo 😄

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp:132]: changed to `EXPECT_*`

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:215]: I changed to EMA on hz (removing `count`) with a heavy weight on new hz (0.9*new_hz + 0.1*old_hz) and it works, but I'm thinking this way I would I also need to detect a step change in message rates (in the event someone dynamically modifies pub rate) to reset the EMA.

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:125]: removed

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:272]: removed

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp:171]: I pushed it to a Test Fixture SetUp.

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp:186]: :+1:

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:215]: I added another test to capture behavior of the policy and the EMA when pivot message changes rate and new pivot is selected.

Also, I moved `compute_h` to the `Rate` class.

File: test/test_latest_time_policy.cpp
Comment (chapulina) [inline test/test_latest_time_policy.cpp:236]: I'm a bit confused about this test case. If I understand correctly, all 3 (p, q and r) have new values at `idx==7`. But that's not being published until `idx==9`. 

I understood that the filter was supposed to publish at the fastest rate possible, whenever there's a new message. So I'd expect it to publish at `idx==5` (when q changes), and again at `idx==7` (when they all change), so the subscriber always has the most up-to-date information.

Is my assumption correct?

Comment (andermi) [reply test/test_latest_time_policy.cpp:236]: I could try to change the behavior, but the last computed rate for `p` is 50Hz until it is received again when it is received at `idx==7` but it is received after `q` at `idx==7`. `q` would be the new pivot, but the policy doesn't know that yet until `q` is received at `idx==9`. I could change this to detect that `p` is late, but I'd have to think about side-effects and what to do if it's late.

Comment (andermi) [reply test/test_latest_time_policy.cpp:236]: > I'd expect it to publish at `idx==5` (when q changes)

The computed rate for `p` is still 50Hz at `idx==5` since it was computed as 50Hz at `idx==4` and won't be computed again until `idx==7`.

Comment (andermi) [reply test/test_latest_time_policy.cpp]: I now loop over the sorted rates and pick the first message that isn't late as the pivot. This has the effect of publishing on `idx==7`

Comment (andermi) [reply test/test_latest_time_policy.cpp]: Basically, q now detects that p is late at idx==7, but can't at idx==5 since p still has a chance to arrive after q and still be on time.

Comment (andermi) [reply test/test_latest_time_policy.cpp]: > publish at the fastest rate possible, whenever there's a new message

This would have the effect of publishing at a fairly erratic rate. For example, if `p` and `q` were both 50Hz, we'd get two publishes of all messages. If `p` and `q` came in 0.01s apart, everything would look like a ZOH @ 100Hz. If they came in closer/further apart, the publishing would be galloping. The publishing would be much more erratic for more messages.

Comment (chapulina) [reply test/test_latest_time_policy.cpp]: Thanks for the explanation, the consequences of publishing all the time are clear to me now 👍🏽

**Review by chapulina — APPROVED** [review]
Thanks for iterating, LGTM! I'd appreciate a review from one of the maintainers though.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:95]: What's this unused argument for?  Can we just remove it?

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:191]: I don't love the fact that this is a recursive function.  I guess in the nominal case it should only ever recurse once, but it makes it harder to reason about.  Could we possibly make this a loop instead?

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:196]: Similarly, where do these magic numbers come from, and should they be configurable?

**Review by clalancette — CHANGES_REQUESTED** [review]
Overall, looks pretty good to me.  I've left a few things inline that I think we should address.

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:95]: yep! removed

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:191]: turned it into a do-while

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:187]: config now

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:196]: also configs

**Comment by andermi** [comment]
@clalancette I think/hope I've addressed your comments.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:172]: Please remove this debugging statement.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:175]: Ditto.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:179]: Ditto.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:251]: Ditto.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:264]: I'm not a huge fan of this.  This resolves into an `assert` statement, which means that when building in release mode, this protection won't be in place.  I don't think this is a very performance-sensitive piece of code, so I'd rather that we just make this an if statement that exits the function if we see it.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:341]: Please remove this debugging statement.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:346]: Please remove this debugging statement.

File: include/message_filters/sync_policies/latest_time.h
Comment (clalancette) [inline include/message_filters/sync_policies/latest_time.h:77]: Given what is actually used in this header file, I think this whole block can be reduced to:

```
#include <algorithm>
#include <mutex>
#include <numeric>
#include <tuple>
#include <vector>
```

File: test/test_latest_time_policy.cpp
Comment (clalancette) [inline test/test_latest_time_policy.cpp:36]: Add:

```
#include <functional>
#include <memory>
#include <vector>
```

**Review by clalancette — CHANGES_REQUESTED** [review]
All right, one last pass.  Once these things are fixed, we'll be ready to run CI on it.

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:172]: sorry about all the debugs :disappointed: 
removed!

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:264]: done

File: include/message_filters/sync_policies/latest_time.h
Comment (andermi) [reply include/message_filters/sync_policies/latest_time.h:77]: done

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp:36]: done

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16803/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11382/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17208/)

**Comment by andermi** [comment]
I actually did find a small intermittent bug

**Comment by clalancette** [comment]
Let's run Windows CI again to see where we are at: * Windows [Build Status](http://ci.ros2.org/job/ci_windows/17217/)

File: test/test_latest_time_policy.cpp
Comment (chapulina) [inline test/test_latest_time_policy.cpp:322]: Instead of just sleeping a fixed amount of time between loops, is there a condition that could be checked here so it waits until that's satisfied? That would remove the reliance on time and make the test robust on slower machines.

One common pattern I've used is something like this:

```cpp
int sleep{0};
int maxSleep{30};
for (; condition && sleep < maxSleep; ++sleep)
{
  // sleep
}
EXPECT_LT(sleep, maxSleep);
EXPECT_TRUE(condition);
```

**Review by chapulina — COMMENTED** [review]
It looks like there are some Windows warnings being introduced by this PR:

https://ci.ros2.org/job/ci_windows/17217/msbuild/new/#issuesContent

File: test/test_latest_time_policy.cpp
Comment (andermi) [reply test/test_latest_time_policy.cpp:322]: Well, the rate sleep is for creating message rates.

I suppose instead of using a rate sleep, I could pass the `Policy` an `rclcpp::Clock` that is set for sim time, and control the `/clock` in the test? Then the time the messages are received would be deterministic.

Requiring stamped messages could also resolve timing issues. I was hoping to make it generic, but users should really be using stamps for messages that require timing anyways :stuck_out_tongue_winking_eye:

**Comment by quarkytale** [comment]
Retriggered CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16963/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11527/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17362/)

**Comment by quarkytale** [comment]
This is currently blocked by timing issue in windows CI.
There are two EMA filters tuned for Linux: 
1. mean rate for incoming message
2. mean error from 1 which is similar to stddev 

In Linux, a threshold of 10 * EMA#2 is used for detecting if the message rate changed. On windows, the rate change is not being detected the same way.

**Comment by andermi** [comment]
@clalancette I gave it another shot... I spun up ros2 dev on my windows machine.
The test passes using sim /clock to control the test timing

**Comment by quarkytale** [comment]
Retriggered Windows CI [Build Status](https://ci.ros2.org/job/ci_windows/17557/)

**Comment by andermi** [comment]
Looks like compiler warnings causing `unstable`. Working through them now.

**Comment by andermi** [comment]
@quarkytale would you mind running another one? I fixed the 7 msbuild warnings related to conversion from `size_t`. Can you see anything else to fix first?

**Comment by quarkytale** [comment]
I'm not sure if the warnings were an issue, but good to have them cleared. Seems like some tests are being skipped, that might be the reason for unstable status. Other things that stood out to me:
- `[WARNING] [python.exe-3]: 'SIGINT' sent to process[python.exe-3] not supported on Windows, escalating to 'SIGTERM'`
- `[-ERROR-] IO exception has been thrown: java.nio.file.NoSuchFileException: C:\ci\ws\src\ros2\message_filters\include\message_filters\sync_policies\latest_time.h`

Let's see if any of these persist: [Build Status](https://ci.ros2.org/job/ci_windows/17558/)

**Review by quarkytale — APPROVED** [review]
LGTM and CI is green!

**Comment by adityapande-1995** [comment]
https://github.com/Mergifyio backport humble

**Pull request merged**
